### PR TITLE
[API] Limit the number of submissions done

### LINKED ISSUE #717: [API] Limit the number of submissions
- [ ] Extend the api to store the maximum number of submissions
- [x] Check that the number of submissions has not been exceeded and the user is still allowed to make further submissions

 --- 
ref #716

### PR DESCRIPTION
**Pull request by SebastianEngl**
resolves #717

### CODE REVIEW

File: modules/fbs-core/api/src/main/resources/migrations/14_sql_count_submission.sql [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+if anzahl > maxAnzahl + 1 then
+ signal sqlstate '45000';
+ set new.val = NULL;
+END IF;
+END;
+DELIMITER ;
+
+
+INSERT INTO migration (number) VALUES (12);
```
Comment (pdrtzk) [inline modules/fbs-core/api/src/main/resources/migrations/14_sql_count_submission.sql:37]:
```suggestion
INSERT INTO migration (number) VALUES (14);
```
The migration number in this statement must be the same as the number used in the file name. 
```suggestion
INSERT INTO migration (number) VALUES (12);
```
Thats  also why the pipline check for migrations fails.

File: modules/fbs-core/api/src/main/resources/migrations/14_sql_count_submission.sql [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+;
+
+if anzahl > maxAnzahl + 1 then
+ signal sqlstate '45000';
+ set new.val = NULL;
+END IF;
+END;
+DELIMITER ;
+
+
```
Comment (mxsph) [inline modules/fbs-core/api/src/main/resources/migrations/14_sql_count_submission.sql:36]:
```suggestion
BEGIN;
CREATE TRIGGER checkAttempts before insert on user_task_submission FOR EACH ROW BEGIN
    set @attempts = (SELECT count(submission_id)
          FROM user_task_submission
          WHERE user_id = New.user_id AND task_id = New.task_id);

    set @maxAttempts = (SELECT CASE WHEN attempts IS NULL THEN -1 ELSE attempts END
    FROM task
    WHERE task_id = NEW.task_id);

    IF @maxAttempts != -1 AND @attempts + 1 > @maxAttempts THEN
     signal sqlstate '45000';
    END IF;
END;

ALTER TABLE task ADD COLUMN attempts INTEGER;

```

I would still make some improvements to the trigger (see diff):
  - [ ] if maxAttempts is not set it should be ignored
  - [ ] `DELIMITER` is not needet in jdbc
  - [ ] `submission` is the wrong table (it should be `user_task_submission`
  - [ ] the SQL queries can be truncated

**Review by mxsph — CHANGES_REQUESTED** [review]

**Review by mxsph — APPROVED** [review]

**Review by pdrtzk — DISMISSED** [review]

File: modules/fbs-core/api/src/main/resources/migrations/14_sql_count_submission.sql [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
```
Comment (Zitrone44) [inline modules/fbs-core/api/src/main/resources/migrations/14_sql_count_submission.sql:1]:
```suggestion
```

File: modules/fbs-core/api/src/main/resources/migrations/14_sql_count_submission.sql [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    IF @maxAttempts != -1 AND @attempts + 1 > @maxAttempts THEN
+     signal sqlstate '45000';
+    END IF;
+END;
+
+ALTER TABLE task ADD COLUMN attempts INTEGER;
+
+INSERT INTO migration (number) VALUES (14);
+
+COMMIT;
```
Comment (Zitrone44) [inline modules/fbs-core/api/src/main/resources/migrations/14_sql_count_submission.sql:21]:
```suggestion
COMMIT;

```

**Review by Zitrone44 — CHANGES_REQUESTED** [review]
Works very well :+1:
Just a minor style fix.

**Review by Zitrone44 — APPROVED** [review]

**Pull request merged**